### PR TITLE
Migrate pallet-balances to pallet attribute macro

### PR DESCRIPTION
**Pull request by ascjones**
Part of #7882.

Converts the `Balances` pallet to the new pallet attribute macro introduced in #6877.

Following the upgrade guidelines here: https://crates.parity.io/frame_support/attr.pallet.html#upgrade-guidelines.

## :warning: Breaking Change :warning:

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storages now use PalletInfo for module_prefix instead of the one given to decl_storage: Thus any use of this pallet in construct_runtime! should be careful to update name in order not to break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the `Balances` pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the `Balances` pallet.

polkadot companion: paritytech/polkadot#2331

### CODE REVIEW

File: frame/balances/src/lib.rs
Comment (thiolliere) [inline frame/balances/src/lib.rs:467]: pallet balances is used by many other pallets, so I guess their tests will fail to find `assimilate_storage`. Either fix the tests to import `GenesisBuild` or do same as you did in frame_system (I.e. implement them manually again).

Comment (ascjones) [reply frame/balances/src/lib.rs:469]: You beat the compiler to telling me this :crab:

File: frame/system/src/lib.rs
Comment (dvdplm) [inline frame/system/src/lib.rs:280]:
```suggestion
		/// A dispatch that will fill the block weight up to the given ratio.
```

File: frame/balances/src/lib.rs
Comment (athei) [inline frame/balances/src/lib.rs:225]: Is this empty block necessary or could we just remove it?

File: frame/balances/src/lib.rs
Comment (athei) [inline frame/balances/src/lib.rs:620]: Does this comment still make sense? Does not seem private here.

**Review by athei — APPROVED** [review]
LGTM. We need a polkadot companion, though.

File: frame/balances/src/lib.rs
Comment (dvdplm) [inline frame/balances/src/lib.rs:290]: Do we have any preference between `pub(crate)` and `pub(super)` for methods that were private before?

File: frame/balances/src/lib.rs
Comment (ascjones) [reply frame/balances/src/lib.rs:225]: It's mandatory https://crates.parity.io/frame_support/attr.pallet.html#hooks-pallethooks-mandatory

File: frame/balances/src/lib.rs
Comment (ascjones) [reply frame/balances/src/lib.rs]: Although they have the same effect in this case, I suppose `pub(super)` is better because it's more restrictive. We should consider updating the migration guide to specify what to prefer. This visibility is required just for tests I assume.

File: frame/balances/src/lib.rs
Comment (dvdplm) [inline frame/balances/src/lib.rs:24]: This link is not right. Maybe all of these can just be `[Thing]` now?

Comment (thiolliere) [reply frame/balances/src/lib.rs:24]: yeah all these "old way" links could be updated.

**Comment by ascjones** [comment]
Companion PR needs approving for CI to pass: https://github.com/paritytech/polkadot/pull/2331

**Comment by ascjones** [comment]
Just waiting for a second approval, @thiolliere @dvdplm

**Review by dvdplm — APPROVED** [review]
LGTM but I'd feel more comfortable if @thiolliere gave thumbs up as well, this being a pretty important pallet...

**Comment by thiolliere** [comment]
i haven't review yet, but would have prefered to fix https://github.com/paritytech/substrate/issues/7949 first, as with the incoming next release of substrate, user pallets can start failing due to their pallet info being `()`, and I can expect some of them to get very confused/lost.
I know none of substrate pallets would but still.

**Comment by ascjones** [comment]
Ok we will wait until #7949 before merging (is there an appropriate label for that?). I'll migrate some more tests this afternoon.

**Comment by thiolliere** [comment]
on-ice can be appopriate

File: frame/balances/src/lib.rs
Comment (gavofyork) [inline frame/balances/src/lib.rs:219]: @thiolliere this feels a bit boilerplatey. can we have the macro automatically do the `PhantomData` dance if we just use a `_` here?

```suggestion
	pub struct Pallet<T, I=()>(_);
```

Comment (ascjones) [reply frame/balances/src/lib.rs:219]: On the other hand it would go against the idea of the new syntax being valid Rust.

Comment (thiolliere) [reply frame/balances/src/lib.rs:219]: yes, here is the PR https://github.com/paritytech/substrate/pull/8091

**Comment by ascjones** [comment]
https://github.com/paritytech/substrate/issues/7949 now done, so ready for a look @thiolliere

File: frame/balances/src/lib.rs
Comment (thiolliere) [reply frame/balances/src/lib.rs:219]: > On the other hand it would go against the idea of the new syntax being valid Rust.

it is true, but at least this syntax is accepted by `syn` crate thus we have a relatively good syntax parser.
Hopefully, the expansion is documented enough not to confuse too much.
And it is same logic as in `#[pallet::storage]`: `_` is a type to be replaced by macro.

**Review by thiolliere — APPROVED** [review]

**Comment by thiolliere** [comment]
bot merge

**Pull request merged**
